### ISSUE TITLE
4-step alternating training

### ISSUE THREAD

**Issue by deepaklunked** [Abhijit-2592/Keras_object_detection#4]
@Abhijit-2592 
Can you explain a bit more on how to use the 4-step alternating training.? I would like to learn more.

**Comment by Abhijit-2592** [comment]
Consider the faster rcnn architecture as containing 3 unique layers. The base-nn, rpn-layer and the final classification layer respectively. 4-step alternating training is essentially freezing and training the above 3 layers in any combination. Refer the paper for more info. 

Here I have hacked around the code to implement it. 
Quote from the train_frcnn.py documentation

> Performing alternating training:
>     - Use the train_rpn,train_final_classifier and train_base_nn boolean arguments to accomplish
>     alternating training.
>     - While using the above arguments change the members of losses_to_watch = ['rpn_cls','rpn_reg','final_cls','final_reg']
>       accordingly else it will throw error
>     - for eg if you are training only the base_nn and the rpn set:
>          train_rpn = True
>          train_base_nn = True
>          train_final_classifier = False
>          losses_to_watch = ['rpn_cls','rpn_reg'] (do not include 'final_cls', 'final_reg')

The train_rpn, train_base_nn, train_final_classifier boolean args are pretty straight forward. The losses_to_watch might be a bit confusing. when I tried freezing the layers of RPN and base_nn in the code from yhenon's repo I was observing changes in the rpn_Regression and rpn_classification losses (Technically this should not have happened). And since the final model is saved on the sum of losses it was creating a problem. Thus I have hacked around a bit by considering only the losses of the layers which are being trained. By layers I mean the base_nn, RPN, Final_classification (not in the meaning of keras layers). Thus if you are freezing the Final_classification, training the RPN and the base_nn only then pass the losses to watch as ['rpn_cls','rpn_reg']

**Comment by ant1pink** [comment]
Why do you freeze a layer during training? Is that because you are satisfied how it is now and you don't want to tune it anymore?

**Comment by Abhijit-2592** [comment]
@ant1pink  refer the official faster rcnn [paper](https://arxiv.org/abs/1506.01497) for more info. 

Naively yes, But the answer is not as direct as that. Practically I have observed a slightly higher performance of 4 step alternative training compared to the joint training proposed in tensorflow object detection API

**Comment by ant1pink** [comment]
@Abhijit-2592 Thanks, I will have a try. 
For my case, the false positive is quite high at the moment. Do you have any suggestions? 
Have you tried RetinaNet yet, is it better?

**Comment by deepaklunked** [comment]
Same here, @ant1pink. I too get a lot of false positives. I let it train with inceptionv3 for some 160 epochs, and the loss seemed pretty low, like ~0.17. Any suggestions, @Abhijit-2592 on how to reduce false positives?

**Comment by ant1pink** [comment]
@deepaklunked Maybe use some dropout layers

**Comment by Abhijit-2592** [comment]
I am closing this issue. Please feel free to open other issues if there are any. Thanks

**Comment by Abhijit-2592** [comment]
Mmm also try increasing the ratio of negative samples in the training. By default the positive and negative samples are in ratio 1:1 try 1:3 or 1:2 this should significantly reduce false positives. This is not implemented, you have to tweak the code slightly to make the ratio 1:3 or 1:2.
